### PR TITLE
Port lightning corrected

### PR DESCRIPTION
**Pull request by szmazurek**
Fixes #

## Proposed Changes
- added lightning module with basic initialization options
- refactored scheduler, optimizer, and metric creational functions
- added unit tests for the newly created primitives
- added new unit tests to CI pipeline
- added usage of new loss interfaces
- this is another issue opened as the previous git history got messed

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].
- [ ] The `logging` library is being used and no `print` statements are left.

### CODE REVIEW

File: GANDLF/losses/__init__.py [hunk 1]
```diff
@@ -38,3 +37,26 @@
...
+    Args:
+        params (dict): The parameters' dictionary.
+
+    Returns:
+        loss (object): The loss definition.
+    """
+    # TODO This check looks like legacy code, should we have it?
+
+    if isinstance(params["loss_function"], dict):
+        chosen_loss = list(params["loss_function"].keys())[0].lower()
```
Comment (sarthakpati) [inline GANDLF/losses/__init__.py:55]: Is this step going to be become redundant after the port?

File: GANDLF/metrics/__init__.py [hunk 2]
```diff
@@ -102,6 +102,26 @@
...
+
+    Returns:
+        metric_calculators (dict): A dictionary containing the calculators of the specified metric functions.
+    """
+    metric_calculators = {}
+    for metric_name in params["metrics"]:
+        metric_name = metric_name.lower()
+        assert (
+            metric_name in global_metrics_dict
+        ), f"Could not find the requested metric '{metric_name}'"
```
Comment (sarthakpati) [inline GANDLF/metrics/__init__.py]: I had a thought: instead of putting this as an assert, can we put this as a `warning` and just output the other metrics that are present (something like [this](https://github.com/szmazurek/GaNDLF/blob/abbd1633e4f6537719e43f86d8543528703fe8b2/GANDLF/compute/loss_and_metric.py#L139))? This would be potentially less obtrusive to a user's workflow.

File: GANDLF/metrics/metric_calculators.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        pass
+
+
+class MetricCalculatorSDNet(AbstractMetricCalculator):
+    def __init__(self, params):
+        super().__init__(params)
+
+    def __call__(self, prediction: torch.Tensor, target: torch.Tensor, *args):
+        metric_results = {}
+        # TODO what do we do with edge case in GaNDLF/GANDLF/compute/loss_and_metric.py?
```
Comment (sarthakpati) [inline GANDLF/metrics/metric_calculators.py]: Which edge case is that?

File: GANDLF/models/lightning_module.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        self._initialize_preds_target_processor()
+
+    def _initialize_model(self):
+        self.model = get_model(self.params)
+
+    def _initialize_loss(self):
+        self.loss = LossCalculatorFactory(self.params).get_loss_calculator()
+
+    def _initialize_metric_calculators(self):
+        # TODO can we have situation that metrics are empty?
```
Comment (sarthakpati) [inline GANDLF/models/lightning_module.py]: I think metrics should never be empty, and this should be a case for assert.

File: GANDLF/losses/__init__.py (hunk 1 above)
Comment (szmazurek) [reply GANDLF/losses/__init__.py:55]: Perhaps, I am not fully aware of that functionality and which config convention does it adhere to, but maybe we can think of standardizing the dict format for loss config. I mean, if the loss_function is indeed dict, then we require it to be something like this: 
```yaml
- "loss_function":
    - "name" : <loss_name>,
    - "config" : <dict_with_custom_loss_config>
````
That would allow us to use the new refactored losses in clean way, thoughts?

File: GANDLF/metrics/__init__.py (hunk 2 above)
Comment (szmazurek) [reply GANDLF/metrics/__init__.py]: Yeah, we can go this way. But on the other hand, maybe the user really wanted to have the specific metric, so early fail would notify about its lack at the very beginning. Or, staying with the assertions, we could move it to a very early stage of the pipeline, when the config is being checked (if the pydantic port happens perhaps the problem will be solved). Anyway, just my speculations, if you think that it should go with the warning then let's do it that way.

File: GANDLF/models/lightning_module.py (hunk 4 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: Ok, will remove the todo then.

File: GANDLF/metrics/metric_calculators.py (hunk 3 above)
Comment (szmazurek) [reply GANDLF/metrics/metric_calculators.py]: I meant this one https://github.com/szmazurek/GaNDLF/blob/53d38491abd4a02b7ff98373b1b549923f3282c2/GANDLF/compute/loss_and_metric.py#L40

File: GANDLF/losses/__init__.py (hunk 1 above)
Comment (sarthakpati) [reply GANDLF/losses/__init__.py:55]: Agreed - should this be a separate PR to the master?

File: GANDLF/metrics/__init__.py (hunk 2 above)
Comment (sarthakpati) [reply GANDLF/metrics/__init__.py]: > pydantic port happens perhaps the problem will be solved

This is a good idea for us to put on that PR, can you please leave a comment?

Let's proceed with a `warning` here to make it consistent.

File: GANDLF/metrics/metric_calculators.py (hunk 3 above)
Comment (sarthakpati) [reply GANDLF/metrics/metric_calculators.py]: Got it. I think this was referring to the implementation in a part of the code that no longer exists, we might not need all this wrapper code.

Also, can you change "age" in that comment line to "edge"? 😆

Comment (szmazurek) [reply GANDLF/metrics/metric_calculators.py]: Sure, did not see that one :P Sure thing

File: GANDLF/metrics/__init__.py (hunk 2 above)
Comment (szmazurek) [reply GANDLF/metrics/__init__.py]: Sure thing, noted

File: GANDLF/losses/__init__.py (hunk 1 above)
Comment (szmazurek) [reply GANDLF/losses/__init__.py:55]: Yeah, can be. I can pack it with other warnings corrections.

Comment (szmazurek) [reply GANDLF/losses/__init__.py:55]: Actually, I see that at master branch, the loss is retrieved in the code, there is not "getter" function, so perhaps we can wait with that change when the lightning will be integrated?

Comment (sarthakpati) [reply GANDLF/losses/__init__.py:55]: Okay, sure.

**Review by sarthakpati — APPROVED** [review]
